homebrew: create selfdestruct cmd

DomT4 opened the pull request:
Previously: #31860
Related: #37667

DomT4 commented:
CC @Homebrew/owners 

We did this 6 months ago, but I think I'm slightly less of an idiot these days and can actually work with your feedback a little more. I rewrote half the script on this, and made it a lot more minimal. Not perfect by any means, but it's better than it was... I think.

DomT4 commented:
```
bash-3.2$ git init
Initialized empty Git repository in /Users/Dominyk/tmpbrew/homebrew/.git/
bash-3.2$ git remote add origin https://github.com/Homebrew/homebrew.git
bash-3.2$ git fetch origin; git reset --hard origin/master
HEAD is now at 5b29475 x264: update r2533 bottle.
=====

env | grep brew
HOMEBREW_CACHE=/Users/Dominyk/tmpbrew/homebrew/brew_cache
HOMEBREW_LOGS=/Users/Dominyk/tmpbrew/homebrew/brew_logs
=====

brew install libsvg
🍺  /Users/Dominyk/tmpbrew/homebrew/Cellar/libpng/1.6.16: 17 files, 1.3M
🍺  /Users/Dominyk/tmpbrew/homebrew/Cellar/jpeg/8d: 18 files, 776K
🍺  /Users/Dominyk/tmpbrew/homebrew/Cellar/libsvg/0.1.4_1: 11 files, 320K
=====

brew --selfdestruct
Homebrew found at: /Users/Dominyk/tmpbrew/homebrew
Deleting files...

The following previously installed formulae were removed:
jpeg
libpng
libsvg

The following previously tapped kegs were removed:

Homebrew has been removed from your system. Thanks for brewing with us!
You may also wish to revert your $PATH to its original state
=====

TREE:
/Users/Dominyk/tmpbrew/homebrew
├── Library
├── bin
└── share
    ├── doc
    │   └── homebrew
    └── man
        └── man1
```

xu-cheng commented:
I think it's better that such feature exists as a part of `homebrew/install` rather than a homebrew command. So users can uninstall homebrew by:

```
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/uninstall)"
```

xu-cheng commented on Library/Homebrew/cmd/--selfdestruct.rb:
```diff
@@ -0,0 +1,93 @@
+rescue LoadError
+  onoe "You must run `sudo /System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/gem install trollop`"
+  raise
+end
```
So to uninstall homebrew, we need install a gem? Can you make it dependency free?

DomT4 replied:
Yes, for now. This is 90% because I found trollop the other day and played with it, and really loved the interface. It's clean, easy and reliable. There's probably an internal Ruby requirement we could use if necessary... I'll just need to hunt it down. We could also lessen the permanence of the gem by installing the gem into ` /tmp ` and then telling Brew to look there for it for the duration of the script.

DomT4 replied:
This isn't completely without precedence though. Various aspects of Brew will require you to install x gem for x purpose.

xu-cheng replied:
> This isn't completely without precedence though. Various aspects of Brew will require you to install x gem for x purpose.

But those features aren't for uninstalling something. `ARGV` should be enough in this script.

DomT4 replied:
Aye, not an unfair point. I'm quite fond of the idea of installing the gem into ` /tmp ` for the duration, but I'll shift away entirely if it's deemed too odious.

tdsmith replied:
No new gem deps, please!

DomT4 replied:
@tdsmith Even to ` /tmp `? :angel:

chdiza replied:
> No new gem deps, please!

Amen, regardless of where they're put.

DomT4 replied:
I'll move away from ` trollop `, as nice as it is. I believe ` optparse  ` which this used before is native. It's clunkier, and I'll need to go readup on the syntax a bit, but shouldn't be too hard.

xu-cheng commented on Library/Homebrew/cmd/--selfdestruct.rb:
```diff
@@ -0,0 +1,93 @@
+abort 'Failed to locate any homebrew files!' if $files.empty?
+
+rm =
+  if @options[:dryrun]
```
I prefer an interactive version of uninstall script. Just like what we do in the install script.

DomT4 replied:
Meh. I don't like interactivity, to be honest. Lots of opportunity for things to go wrong. I'm happy to stick an interactive message back towards the top, as I had before, along the lines of ` THIS WILL DESTROY YOUR HOMEBREW. Abort unless /yes/y` or similar - But I don't really want to make it super-interactive. I think we'd be causing maintenance nightmares for ourselves with interactivity.

xu-cheng replied:
I prefer interactive as I treated it as the same of install script. However, this is just my personal opinion. Let's hear other maintainers'  thoughts. :wink:

DomT4 replied:
There's an extremely minimal amount of interactivity in the install script as is. It's pretty much the bare minimum along the lines of ` Hey we're gonna install Homebrew right now, you cool with that? `, which we could do here as well, but I'm not sure what use interactivity would have in an uninstall script beyond that. Happy to take other views though, Don't want to flood the entire thread with our discussion :smile_cat:

tdsmith replied:
I think a "This is what this script will do; type yes to continue:" prompt is probably a good idea just because it's such a destructive command; perhaps `--force` makes it run non-interactively.

DomT4 replied:
That'd be in keeping with the install script check. Makes sense if we have a check to install, It'd be a solid idea to have a check on uninstall as well, especially since as you highlight the last thing you'd want to do is to have someone stumble into the command accidentally or go *I wonder what happens if I use this new command* :smile_cat:. Force being the override sounds reasonable.

DomT4 commented:
Heh, if people already get paranoid about a script that installs Homebrew onto your Mac with the ` -fsSL ` flag I can't imagine response is going to be super-shiny to a script that is actively designed to go through various aspects removing things.

I kept it as a ` cmd ` because doing so means we can tap Homebrew's internal knowledge a lot more, with ` HOMEBREW_PREFIX ` and ` HOMEBREW_CACHE ` and so on, rather than having to look for them or shell out so much. Us knowing where various Homebrew elements were located but shelling out to look for them regardless is what caused a little angst in the last PR. It's reasonably portable though if we go back to shelling out in places.

xu-cheng commented:
> I kept it as a cmd because doing so means we can tap Homebrew's internal knowledge a lot more, with HOMEBREW_PREFIX and HOMEBREW_CACHE and so on

We can achieve the same knowledge from an external script. Like this:
```
require "pathname"
HOMEBREW_LIBRARY_PATH = Pathname.new(`brew --prefix`.strip)/"Library/Homebrew"
$:.unshift HOMEBREW_LIBRARY_PATH
require "global"
```

DomT4 commented:
That's still 4 lines of code in place of 1 line, and includes an element where we have to shell out regardless. Shelling out and multiple lines of code where 1 will suffice is something I was actively trying to avoid on this, just for once... I appreciate my usual style isn't concise :wink:.

tdsmith commented:
Seems like `selfdestruct` should be a verb and not a flag to me, i.e. not `--selfdestruct`.

I don't understand why having to curl-pipe uninstall would be desirable; that symmetry isn't very compelling to me.

DomT4 commented:
> Seems like selfdestruct should be a verb and not a flag to me, i.e. not --selfdestruct.

As in just ` brew selfdestruct ` rather than ` brew --selfdestruct `? No objections from me if so. Makes sense.

tdsmith commented on Library/Homebrew/cmd/--selfdestruct.rb:
```diff
@@ -0,0 +1,93 @@
+  else
+    # Shelling out here enables us to get brew to do the heavy-lifting
+    # around removing now-dead symlinks and directories.
+    `brew list | pbcopy && brew remove $(pbpaste)`
```
should try and avoid using the pasteboard here

xu-cheng replied:
`brew list -1 | xargs brew remove` should work

xu-cheng replied:
Also, why not just deleting the Cellar then `brew prune`?

DomT4 replied:
Normally I wouldn't go near it for this sort of thing, but as a one-liner it's quite useful. Getting this idea of using ` brew ` to do the heavy-lifting in terms of removing formulae prior to removing itself and consequently scooping up all the symlinks and such turned out to be a bit of a PITA to implement. As a shell command inside a wider brew script it threw a tantrum when I tried the easier/nicer `brew remove $(brew list)`. Using the pasteboard was one of the few elements I could get to work 100% of the time without errors. I figured being on the same line/executed at the same time reduced the risk of cross-contamination, but I agree, It's not ideal.

DomT4 replied:
> Also, why not just deleting the Cellar then brew prune?

` brew prune ` can be kind of slow when a large number of formulae are involved. ` brew remove ` seemed quicker by a noticeable degree.

xu-cheng replied:
Under this content, robust is more important than performance. Using pasteboard is a serious security flaw.

DomT4 replied:
That's a tad hyperbolic. There are many things which ` brew ` does that can be construed as security flaws, but a command that exists fleetingly in a controlled environment most users will never stumble onto is minimal at best.

tdsmith replied:
We shouldn't use the pasteboard; that's a really weird side effect for an uninstall script. Using fewer lines of code is a low priority. Slurp the packages into a list and iterate in Ruby if you have to.

DomT4 replied:
I'll probably cave and use ` brew prune `. The performance hit is a nuisance, but I'm not *that* attached to the shell hack. I'll tweak it in a bit. Need to do another round of testing.

xu-cheng commented:
> I don't understand why having to curl-pipe uninstall would be desirable; that symmetry isn't very compelling to me.

To me the reasons are:
* Make it independent with homebrew self. In case a user gets into a unexpected semi-(un)installed state, the curl-pipe will always work. But a brew command may not. 
* Avoid the user accidentally running the uninstall script as it's not preexisting on the system.

tdsmith commented:
Mixing the CLI API (`` `brew list` ``) with the internal API (`HOMEBREW_PREFIX`) is odd and I think we should avoid it; picking one is probably fine. Using the CLI API might be easier to reason about.

This should make sure to use the `brew` it's installed alongside and not the `brew` from $PATH.

Pruning after removing everything is probably a good idea.

tdsmith commented:
Making the script available apart from Homebrew sounds okay but we'd have to have an implementation that doesn't use `brew` or any Homebrew API for that to matter.

DomT4 commented:
> Mixing the CLI API (`brew list`) with the internal API (HOMEBREW_PREFIX) is odd and I think we should avoid it

The larger mix in this PR largely came around because I tried to avoid the situation where we know where HOMEBREW_PREFIX and co are, but we dilly around the system looking for them or shell out regardless. It made sense to use the internal API in that situation, even if it makes the script less portable.

Agree on the ` brew list `. Ideally, I'd like to kill that off and use a chunk of the internal API to generate brew list, but I'm not aware of an easy-easy way to do that. The `brew list` command is incredibly clean, and thorough. Not sure how to internalise that call without ending up replicating vast chunks of the `brew list` cmd.

> This should make sure to use the brew it's installed alongside and not the brew from $PATH.

Theoretically, It should be doing this already, but I'll double check later. Should be doable though, if it isn't done already.

> but we'd have to have an implementation that doesn't use brew or any Homebrew API for that to matter.

The gist this is based on is essentially that. I tore chunks out and put new chunks in, but the core idea is the same. There's a link to the gist at the top of the command here.

chdiza commented:
What about the things brew plops into `/etc` or `/share` that aren't symlinks?  (E.g., a stray `luarocks` folder, or `/etc/openssl`.)  I think gems and eggs might be in a similar boat, so are shell completions.

I fairly regularly destroy my brew installation (manually from terminal, using `brew` and `rm -rv`), and things like that get left behind because there seems to be no way for HB to know that they belong to HB.

DomT4 commented:
> What about the things brew plops into /etc

For now, I'm just somewhat brutally murdering the entire ` etc ` directory on line ~74 or somewhere around there. It seems reasonable to assume it's not necessarily a directory people will have natively, but nonetheless, I'll almost certainly stick an opt-out in the script to allow people to keep their ` etc ` directories as desired.

If everything was stashed inside a named directory in `etc` we could just grab everything from `brew list` and then ` rm_rf ` all folders which shared the same name as an installed formulae, but that has a couple of problems:

1) We can't do anything about old formulae installations long since deleted but the remnants in ` etc ` remain.
2) An awful lot of stuff isn't in named folders, It's directly in ` etc `.

It's a bit of a pain to handle, the ` etc ` directory. Same with ` share `. Share seems to cause less problems in testing, but it's still a small PITA.

Ultimately, I don't think this script *can* end up removing absolutely everything 100% of the time. It will remove almost everything, but things outside the purview of ` brew prune ` or non-symlinks outside the ` HOMEBREW_CELLAR ` sandbox are difficult to interpret and handle correctly.

DomT4 commented:
To keep progress updated, I've pretty much finished addressing the concerns raised here at this point. I'm finishing off the gem-free options, I've moved away from the brief pasteboard interaction in favour of just murdering the Cellar and then pruning, I've added an interactive dire warning about the script's capabilities requiring user acceptance before it proceeds, I've renamed the command so it isn't ` --example ` but rather just ` brew example `, similar to ` brew list ` and co.

I'll shove it through another round of testing in the morning to see it it's in a place where I'm happy to push remote again.

Stuff that still needs working on next:

* Are we happy with just butchering the entire ` etc ` directory, with an optional user override to prevent that from happening.
* Ensure it finds the Brew it has been triggered from, and not just the first Brew that it stumbles across in the ` $PATH `.

xu-cheng commented:
IMHO, I still think it's best that we put this feature in `homebrew/install` repo. Therefore, we can avoid following issues in the future:
* User: My homebrew is broken. How to uninstall it?
* Maintainer: Run `brew selfdestruct`
* User: But my homebrew is broken, I cannot run any brew command.

tdsmith commented:
We'd have to have an implementation that never calls `brew` or uses any homebrew internals for that to matter.

DomT4 commented:
Except if your whole ` brew ` installation is broken, you're screwed anyway:

We can't detect where Brew is, except with very blunt shell methods ` which brew ` and similar, we can't detect what Brew has installed, unless we ` ls -1 ` the whole Cellar, we can't use ` brew ` to do the heavy lifting around removing dead symlinks and folders like ` pkgconfig ` and similar in ` lib `, so we'd have to remove dead symlinks in a more manual way, in a way that could end up removing other dead symlinks in the same directory that the user may prefer we didn't yank, and we'd have to liberally sprinkle ` rm_rf ` around with near reckless abandon.

When we have so much internal information about brew, and can extract more from the ` brew ` command where necessary, it seems self-defeating to use a script that would be inherently *much* more limited in capability, accuracy and information available.

xu-cheng commented:
> We'd have to have an implementation that never calls brew or uses any homebrew internals for that to matter.

:+1: It's doable.

* To locate homebrew, we can try to locate it in usual place then ask user to confirm. If it's not found, we can ask user pass the homebrew place by ARGV.
 * To delete file, we delete Cellar then glob all the dead symlink.

xu-cheng commented:
@DomT4 Using less code to implement such feature is really low priority. What matters is the feature itself and robustness. Normal users won't uninstall homebrew very often. And when they do, it's more than likely that homebrew is broken in some degree.

DomT4 commented:
It's doable, but it isn't smart, and I think an uninstall script that people can curl-pipe in is a pretty awful idea, personally.

By all means, If you want to produce such a script, I can't stop you, and I can't stop the maintainers from accepting it. I am not vested with such power :smile_cat: :smile_cat:. But there are a million scripts out there to remove Brew that don't use any internal information, that don't use the power ` brew ` has, and I don't particularly see the upside of writing a million and one-th script on the same basis.

DomT4 commented:
> Using less code to implement such feature is really low priority.

The particularly point I made above isn't about less code, or more code. Although I consider less code = easier to read = less monumental screw ups/misunderstandings, It's about being smart, using methods ` brew ` already offers, being self-contained, and offering a better experience than the million scripts out there already that outsource from ` brew ` in favour of using less precise measures. It is not a more robust experience to produce a less accurate script that can't tap the self-understanding ` brew ` has. That is my point, not a competition about who has the greater length and whether that length brings great fireworks to the party.

mikemcquaid commented:
> IMHO, I still think it's best that we put this feature in homebrew/install repo.

Personally I agree with this. I'd be the reason a bunch of users want to uninstall is because their Homebrew is broken. We could perhaps use `brew` commands to try and doing a cleaner uninstall or something but not assume any of them would pass.

Sorry to disagree here @DomT4; I do appreciate the hard work that's been put in here.

chdiza commented:
Don't forget to get rid of the following litter: `/usr/local/*.md` and `/usr/local/README.txt`.

DomT4 commented:
I don't have the energy to 'fight' two maintainers on whether this should go in ` Install ` or not just lately, so I'll finish this off locally, stick it in a new tap somewhere, and let Xu deal with the non-internal-knowledge script.

DomT4 commented:
Finally finished tidying up, and making the script execute itself inside ` brew ` without brew wondering what the heck I was talking about. If there's ever any interest from anyone else, it's here now: https://github.com/DomT4/homebrew-removal
